specs: use lazy lexicographic comparison instead of key_ordering

tgamblin opened the pull request:
We have been using the `@llnl.util.lang.key_ordering` decorator for specs and most of their components. This leverages the fact that in Python, tuple comparison is lexicographic. It allows you to implement a `_cmp_key` method on your class, and have `__eq__`, `__lt__`, etc. implemented automatically using that key. For example, you might use tuple keys to implement comparison, e.g.:

```python
class Widget:
    # author implements this
    def _cmp_key(self):
        return (
            self.a,
            self.b,
            (self.c, self.d),
            self.e
        )

    # operators are generated by @key_ordering
    def __eq__(self, other):
        return self._cmp_key() == other._cmp_key()

    def __lt__(self):
        return self._cmp_key() < other._cmp_key()

    # etc.
```

The issue there for simple comparators is that we have to build the tuples *and* we have to generate all the values in them up front. When implementing comparisons for large data structures, this can be costly.

This PR replaces `@key_ordering` with a new decorator, `@lazy_lexicographic_ordering`. Lazy lexicographic comparison maps the tuple comparison shown above to generator functions. Instead of comparing based on pre-constructed tuple keys, users of this decorator can compare using elements from a generator. So, you'd write:

```python
@lazy_lexicographic_ordering
class Widget:
    def _cmp_iter(self):
        yield a
        yield b
        def cd_fun():
            yield c
            yield d
        yield cd_fun
        yield e

    # operators are added by decorator (but are a bit more complex)
```

There are no tuples that have to be pre-constructed, and the generator does not have to complete. Instead of tuples, we simply make functions that lazily yield what would've been in the tuple. If a yielded value is a `callable`, the comparison functions will call it and recursively compare it. The comparator just walks the data structure like you'd expect it to.

The ``@lazy_lexicographic_ordering`` decorator handles the details of implementing comparison operators, and the ``Widget`` implementor only has to worry about writing ``_cmp_iter``, and making sure the elements in it are also comparable.

Using this PR shaves another 1.5 sec off the runtime of `spack buildcache list`, and it also speeds up Spec comparison by about 30%. The runtime improvement comes mostly from two things:

1. lazily stopping the comparison as soon as possible (e.g., many specs just have different names, which is the firs thing the lazy generators return)
2. avoiding the use of `hash()`  in `_cmp_iter()` (it was used in `_cmp_key()` before)

tgamblin committed:
specs: use lazy lexicographic comparison instead of key_ordering

We have been using the `@llnl.util.lang.key_ordering` decorator for specs
and most of their components. This leverages the fact that in Python,
tuple comparison is lexicographic. It allows you to implement a
`_cmp_key` method on your class, and have `__eq__`, `__lt__`, etc.
implemented automatically using that key. For example, you might use
tuple keys to implement comparison, e.g.:

```python
class Widget:
    # author implements this
    def _cmp_key(self):
        return (
            self.a,
            self.b,
            (self.c, self.d),
            self.e
        )

    # operators are generated by @key_ordering
    def __eq__(self, other):
        return self._cmp_key() == other._cmp_key()

    def __lt__(self):
        return self._cmp_key() < other._cmp_key()

    # etc.
```

The issue there for simple comparators is that we have to bulid the
tuples *and* we have to generate all the values in them up front. When
implementing comparisons for large data structures, this can be costly.

This PR replaces `@key_ordering` with a new decorator,
`@lazy_lexicographic_ordering`. Lazy lexicographic comparison maps the
tuple comparison shown above to generator functions. Instead of comparing
based on pre-constructed tuple keys, users of this decorator can compare
using elements from a generator. So, you'd write:

```python
@lazy_lexicographic_ordering
class Widget:
    def _cmp_iter(self):
        yield a
        yield b
        def cd_fun():
            yield c
            yield d
        yield cd_fun
        yield e

    # operators are added by decorator (but are a bit more complex)

There are no tuples that have to be pre-constructed, and the generator
does not have to complete. Instead of tuples, we simply make functions
that lazily yield what would've been in the tuple. If a yielded value is
a `callable`, the comparison functions will call it and recursively
compar it. The comparator just walks the data structure like you'd expect
it to.

The ``@lazy_lexicographic_ordering`` decorator handles the details of
implementing comparison operators, and the ``Widget`` implementor only
has to worry about writing ``_cmp_iter``, and making sure the elements in
it are also comparable.

Using this PR shaves another 1.5 sec off the runtime of `spack buildcache
list`, and it also speeds up Spec comparison by about 30%. The runtime
improvement comes mostly from *not* calling `hash()` `_cmp_iter()`.

alalazo commented on lib/spack/llnl/util/lang.py:
```diff
@@ -227,48 +233,203 @@ def list_modules(directory, **kwargs):
+    return tuple(tuplify(x) if callable(x) else x for x in seq())
+
+
+def doublewrap(f):
```
Minor, but using `decorator` instead of `f` may improve readability:
```suggestion
def doublewrap(decorator):
```

cosmicexplorer replied:
```suggestion
def coerce_decorator_arguments(f):
```
along with @alalazo's suggestion!

tgamblin commented:
Seems there's a bug in here somewhere -- I might not get to fixing that until later today.

alalazo commented on lib/spack/llnl/util/lang.py:
```diff
@@ -291,8 +452,9 @@ def __len__(self):
-    def _cmp_key(self):
-        return tuple(sorted(self.values()))
+    def _cmp_iter(self):
+        for _, v in sorted(self.items()):
```
I assume the ordering is arbitrary for us, but want to note that this is different from the previous one (the ordering is done here on `(key, value)` pairs, while before it was just the values)

alalazo commented on lib/spack/spack/test/spec_dag.py:
```diff
@@ -519,6 +519,11 @@ def test_normalize_mpileaks(self):
         assert spec == expected_flat
         assert spec.eq_dag(expected_flat)
 
+        print(spec.tree(color=True))
+        print(expected_normalized.tree(color=True))
+        print(spec == expected_normalized)
+        print(spec != expected_normalized)
```
Are these debug prints to be removed?

alalazo commented on lib/spack/spack/test/spec_syntax.py:
```diff
@@ -828,4 +828,6 @@ def test_compare_abstract_specs(self):
 
         for a, b in itertools.product(specs, repeat=2):
             # Check that we can compare without raising an error
+            print("a:", str(a), type(a))
+            print("b:", str(b), type(b))
```
Same question as above.

alalazo reviewed (dismissed):
Minor comments from a first read of the PR. I'll test drive it asap.

cosmicexplorer commented on lib/spack/llnl/util/lang.py:
```diff
@@ -227,48 +233,203 @@ def list_modules(directory, **kwargs):
-       ``__gt__``, or ``__ge__`` defined, this decorator will overwrite them.
+    or::
+
+        @decorator
```
```suggestion
        @decorator  # Calls the function with zero arguments.
```

tgamblin replied:
fixed

cosmicexplorer commented on lib/spack/llnl/util/lang.py:
```diff
@@ -227,48 +233,203 @@ def list_modules(directory, **kwargs):
+
+        @decorator
+
+    See https://stackoverflow.com/questions/653368.
```
I would usually make this a comment just below the docstring, as it seems like an implementation detail.

tgamblin replied:
fixed

cosmicexplorer commented on lib/spack/llnl/util/lang.py:
```diff
@@ -227,48 +233,203 @@ def list_modules(directory, **kwargs):
+    tuples. The issue there for simple comparators is that we have to
+    bulid the tuples *and* we have to generate all the values in them up
+    front. When implementing comparisons for large data structures, this
+    can be costly.
```
Should we be using this whenever possible, then? Is there a use case where this doesn't apply?

cosmicexplorer commented on lib/spack/llnl/util/lang.py:
```diff
@@ -227,48 +233,203 @@ def list_modules(directory, **kwargs):
+
+    Note: if a class already has ``__eq__``, ``__ne__``, ``__lt__``,
+    ``__le__``, ``__gt__``, ``__ge__``, or ``__hash__`` defined, this
+    decorator will overwrite them.
```
Could we make it raise a `TypeError` instead, unless a keyword argument is provided to the decorator? I'm concerned silently overwriting methods by default could be difficult to debug at some point.

tgamblin replied:
It's hard to know this b/c every object has these methods:

```python
>>> o = object()
>>> hasattr(o, "__lt__")
True
```

I haven't seen other decorators that go to great lengths to avoid this -- if you use the decorator, it's pretty explicit how comparison should work.

cosmicexplorer commented on lib/spack/llnl/util/lang.py:
```diff
@@ -227,48 +233,203 @@ def list_modules(directory, **kwargs):
+        (s is not o) and (o is None or ne(s._cmp_iter, o._cmp_iter)))
+
+    add_func_to_class(
+        "__lt__", lambda s, o:
```
If we break these `lambda`s out into `def`s and use `@memoized`, does that affect performance? Could it affect performance in the future?

tgamblin replied:
we don't want these to be memoized, as they could be used on objects that change (like specs).  `Spec` specifically memoizes the hash (because we already handle invalidation for that) but I don't think it should be the default.

cosmicexplorer commented on lib/spack/spack/spec.py:
```diff
@@ -4225,6 +4231,16 @@ def target(self):
         return self.architecture.target.microarchitecture
 
+    def __hash__(self):
+        """Memoized version of lazy_lexicographic_ordering hash."""
```
Is there a reason this isn't implemented in the `lazy_lexicographic_ordering` decorator? It seems useful for a lot more than the Spec class.

tgamblin replied:
It is implemented there.  This one is an override.  Note that Spec is defined with:

```python
@lazy_lexicographic_ordering(set_hash=False)
class Spec(...):
...
```

cosmicexplorer commented on lib/spack/llnl/util/lang.py:
```diff
@@ -227,48 +233,203 @@ def list_modules(directory, **kwargs):
-           lambda s, o: o is None or s._cmp_key() >= o._cmp_key())
-
-    setter('__hash__', lambda self: hash(self._cmp_key()))
+    if not has_method(cls, "_cmp_iter"):
```
I might add a TODO here to try to check this with a mypy protocol before runtime (https://mypy.readthedocs.io/en/stable/protocols.html).

becker33 commented on lib/spack/llnl/util/lang.py:
```diff
@@ -227,48 +233,203 @@ def list_modules(directory, **kwargs):
+        return not eq(lseq, rseq)
+
+    def gt(lseq, rseq):
+        return lt(lseq, rseq, operator.gt, right_shorter)
```
```suggestion
        return lt(rseq, lseq)
```
Then we can remove op and diff as arguments to `lt`.

tgamblin replied:
fixed

becker33 commented on lib/spack/llnl/util/lang.py:
```diff
@@ -227,48 +233,203 @@ def list_modules(directory, **kwargs):
+        "__ne__", lambda s, o:
+        (s is not o) and (o is None or ne(s._cmp_iter, o._cmp_iter)))
+
+    add_func_to_class(
```
Add a note that all objects are greater than `None` (using python2 semantics in python3).

tgamblin replied:
I mentioned this in the docstring.

becker33 commented on lib/spack/spack/architecture.py:
```diff
@@ -329,23 +329,27 @@ def __repr__(self):
-
-@key_ordering
+    def _cmp_iter(self):
+        yield self.name
```
All yielded values need to `or` to something sane for python3 comparisons.

```suggestion
        yield self.name or ''
```
etc for all compared values.

tgamblin replied:
I worked the None comparison semantics into lazy lexicographic ordering, so this is actually no longer necessary.  I think this makes our comparisons way easier to write.

becker33 commented on lib/spack/spack/spec.py:
```diff
@@ -686,10 +689,10 @@ def add_type(self, type):
-                self.spec.name if self.spec else None,
-                self.deptypes)
+    def _cmp_iter(self):
+        yield self.parent.name if self.parent else None
```
Can't compare `None` to `str` in python3

tgamblin committed:
specs: speed up traversal by avoiding redundant canonicalization

tgamblin commented:
@alalazo @becker33 @eugeneswalker: this is ready for another look.  With the latest E4S build cache (which has 37k specs), it significantly improves performance for `spack buildcache list`:

`develop`:

```console
       42.68 real        35.00 user         0.75 sys
```

`lazy-lexicographic-spec-comparison`:

```console
       24.71 real        23.80 user         0.46 sys
```

So that shaves > 40% of the time off.

The key is really two things.  First, `__lt__` is faster b/c of the laziness -- specs that are actually less than others bail as early as possible.  For `__eq__`, or rather for specs that are equal, things are trickier, because we do not want `__eq__` to have to traverse an entire spec.  I'm leveraging the dag hash like this:

```python
    def __hash__(self):
        # If the spec is concrete, we leverage the DAG hash and just use
        # a 64-bit prefix of it. The DAG hash has the advantage that it's
        # computed once per concrete spec, and it's saved -- so if we
        # read concrete specs we don't need to recompute the whole hash.
        # This is good for large, unchanging specs.
        if self.concrete:
            if not self._dunder_hash:
                self._dunder_hash = self.dag_hash_bit_prefix(64)
            return self._dunder_hash

        # This is the normal hash for lazy_lexicographic_ordering. It's
        # slow for large specs because it traverses the whole spec graph,
        # so we hope it only runs on abstract specs, which are small.
        return hash(lang.tuplify(self._cmp_iter))
```

So basically, I think this gets us the best of both worlds -- lazy comparison for abstract specs that are small, and leveraging precomputed hashes for specs that are large (which are pretty much all the concrete ones).

tgamblin committed:
specs: remove "or ''" from Spec comparisons

Since `lazy_lexicographic_ordering` handles `None` comparison for us, we
don't need to adjust the spec comparators to return empty strings or
other type-specific empty types. We can just leverage the None-awareness
of `lazy_lexicographic_ordering`.

- [x] remove "or ''" from `_cmp_iter` in `Spec`
- [x] remove setting of `self.namespace` to `''` in `MockPackage`

eugeneswalker reviewed (dismissed):
This looks great! Thank you.

becker33 commented on lib/spack/spack/test/spec_syntax.py:
```diff
@@ -352,10 +352,10 @@ def test_multiple_specs_with_hash(self, database):
         x1 = Spec('a')
         x1.concretize()
-        x1._hash = 'xy'
+        x1._hash = 'xyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy'
```
Why is this necessary?

tgamblin replied:
The fact that `Spec.__hash__` now uses [`self.dag_hash_bit_prefix(64)`](https://github.com/spack/spack/pull/21618/files#diff-9e28cdd6f9cd66e61101de6e9d22b1c33561959e4305c63147bbfbf7b22da936R4342-R4345) means that there need to be at least 64 bits of hash data to get the prefix.  That's 13 base32 characters, so I just went ahead and made these hashes full length.  Every real hash will be 32 characters long (160 bits for SHA1) so I made them proper stand-ins.

becker33 commented on lib/spack/llnl/util/lang.py:
```diff
@@ -227,48 +232,224 @@ def list_modules(directory, **kwargs):
+    def h(self):
+        return hash(tuplify(self._cmp_iter))
+
+    # lambdas below handle special cases with None, then delegate to the
+    # recursive functions that take generators above.
```
This comment is no longer accurate

tgamblin replied:
fixed

becker33 reviewed (changes requested):
Only request is removing a vestigial comment. Otherwise LGTM

tgamblin commented:
@becker33:comments addressed

tgamblin commented:
@alalazo FYI

alalazo commented:
Thanks I'll update #21683
